Sources_v3/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware.swift

```diff
@@ -0,0 +1,56 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// A middleware which updates a channel's read events as websocket events arrive.
+struct ChannelReadUpdaterMiddleware<ExtraData: ExtraDataTypes>: EventMiddleware {
+    let database: DatabaseContainer
+    
+    func handle(event: Event, completion: @escaping (Event?) -> Void) {
+        defer { completion(event) }
+        
+        if let event = event as? MessageReadEvent<ExtraData> {
+            updateReadEvent(for: event.cid, userId: event.userId, lastReadAt: event.readAt)
+            return
+        }
+        
+        if let event = event as? NotificationMarkReadEvent<ExtraData> {
+            updateReadEvent(for: event.cid, userId: event.userId, lastReadAt: event.readAt)
+            return
+        }
+        
+        if let event = event as? NotificationMarkAllReadEvent<ExtraData> {
+            let readsForUser = ChannelReadDTO.load(userId: event.userId, context: database.viewContext)
+            database.write({ _ in
+                readsForUser.forEach { read in
+                    read.lastReadAt = event.readAt
+                    read.unreadMessageCount = 0
+                }
+            }, completion: { error in
+                if let error = error {
+                    log.error("Failed to update channel reads for userId \(event.userId), error: \(error)")
+                }
+            })
+            return
+        }
+    }
+    
+    private func updateReadEvent(for cid: ChannelId, userId: UserId, lastReadAt: Date) {
+        let channel = ChannelDTO.load(
+            cid: cid,
+            context: database.viewContext
+        )
```

Comment:
You shouldn't be touching `viewContext` here at all. This must happen inside the `write` block:
```
database.write { 
   // get the channel
   // get its reads and mutate them
}
```
The channel object you get from the view context can be different than the one you mutate inside `write { }`.

Reply:
Fixed in 60eb7b192816dd80a24b5eb9b8ecae1b9062a851